src/nspawn/nspawn.c

```diff
@@ -2594,9 +2594,15 @@ static int inner_child(
                 return -ESRCH;
         }
 
-        r = mount_systemd_cgroup_writable("", arg_unified_cgroup_hierarchy);
-        if (r < 0)
-                return r;
+	if (cg_ns_supported()) {
+		r = unshare(CLONE_NEWCGROUP);
+		if (r < 0)
```

Comment:
On Sat, Jun 25, 2016 at 12:24:00AM -0700, Evgeny Vereshchagin wrote:
> > @@ -2594,9 +2594,15 @@ static int inner_child(
> >                  return -ESRCH;
> >          }
> >  
> > -        r = mount_systemd_cgroup_writable("", arg_unified_cgroup_hierarchy);
> > -        if (r < 0)
> > -                return r;
> > +	if (cg_ns_supported()) {
> > +		r = unshare(CLONE_NEWCGROUP);
> > +		if (r < 0)
> 
> Well, `systemd-nspawn` doesn't fail on startup. But this breaks `UNIFIED_CGROUP_HIERARCHY`:
This is an indirect consequence of cgroup namespaces. With cgroup namespaces the
container will mount the cgroupfs itself. Hence, mounting the cgroupfs is the
task of systemd *inside* the container as opposed to bind-mount magic when
cgroup namespaces are not available. If we want systemd inside the container to
mount the unified cgroup hierarchy the simplest solution is to pass
`systemd.unified_cgroup_hierarchy=1` as argument to `systemd-nspawn`:

```
systemd-nspawn -D /some/rootfs -b 'systemd.unified_cgroup_hierarchy=1'
```

To be backwards compatible with prior `systemd-nspawn` versions that allow
setting the `UNIFIED_CGROUP_HIERARCHY` env variable we can simply append
`systemd.unified_cgroup_hierarchy=1`. However, when the user simply wants a
shell inside the container things get more complicated since there is no
systemd/init process that sets up the cgroupfs.

Minor point: Note also, that the systemd `v230` release notes state that booting
unified cgroups with kernels `>= 4.5` requires systemd `v230`. This is why I
had trouble using unified cgroups:

	"WARNING: it is not possible to use previous systemd versions with
	systemd.unified_cgroup_hierarchy=1 and the new kernel. Therefore it is
	necessary to also update systemd in the initramfs if using the unified
	hierarchy. An updated SELinux policy is also required."
	(https://lists.freedesktop.org/archives/systemd-devel/2016-May/036583.html)

Since the cgroup namespaces patch here requires that systemd *inside* the
container mounts the cgroup it means that systemd `v230` is required inside the
container with a kernel `>=4.5`.

> 
> > nspawn understands the $UNIFIED_CGROUP_HIERARCHY
>           environment variable to individually select the hierarchy to
>           use for executed containers. By default, nspawn will use the
>           unified hierarchy for the containers if the host uses the
>           unified hierarchy, and the legacy hierarchy otherwise.
> 
> ```sh
> -bash-4.3# grep cgroup /proc/self/mounts
> tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,mode=755 0 0
> cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/usr/lib/systemd/systemd-cgroups-agent,name=systemd 0 0
> cgroup /sys/fs/cgroup/net_cls cgroup rw,nosuid,nodev,noexec,relatime,net_cls 0 0
> cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
> cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
> cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
> cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
> cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
> cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
> cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset 0 0
> 
> -bash-4.3# UNIFIED_CGROUP_HIERARCHY=yes systemd-nspawn -D /nspawn-root/ -b 3
> ...
> container# grep cgroup /proc/self/mounts
> tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,mode=755 0 0
> cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/usr/lib/systemd/systemd-cgroups-agent,name=systemd 0 0
> cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
> cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
> cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
> cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset 0 0
> cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
> cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
> cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
> cgroup /sys/fs/cgroup/net_cls cgroup rw,nosuid,nodev,noexec,relatime,net_cls 0 0
> 
> ---
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/systemd/systemd/pull/3589/files/dd8e1b4bf0b4e6180812428053d6dfb97d66b4db#r68485591